Update instructions for running from source + docker

eelanagaraj opened the pull request:
First pass at updating the README -- still needs some work, but most of the very out-of-date sections have been removed and the instructions for running from source/docker have been updated and should run as expected.

barbaraliau commented on README.md:
```diff
@@ -64,40 +65,102 @@ Flags:
+### Version 1: Running from `rosetta` source code
 
-Within the docker image, we pack `rosetta` binary and also `geth` binary from celo-blockchain. Rosetta will run both.
+You will need the following three repositories cloned locally: `rosetta` (this repo), `celo-monorepo`, `celo-blockchain`. You also need: `go >= 1.14`, `rust >= 1.41.0` (blockchain dependency)
```
Can you add links to these repos?

eelanagaraj replied:
good call!

barbaraliau commented on README.md:
```diff
@@ -64,40 +65,102 @@ Flags:
+Prerequisites:
+
+* Checkout `celo-monorepo` branch `alfajores` and run `yarn && yarn build`
+* Checkout `celo-blockchain` tag `v1.1.1` (NOTE: check that this matches the version specified in the `rosetta` `go.mod` file) and `make all`
```
No need to change this now, but making a mental note that this needs to be updated every time the blockchain is upgraded.

eelanagaraj replied:
Yes feel you -- I think this is something we can bundle with streamlining the release process as well, maybe a makefile command/something to just cause all of this to happen with a single command + single place where the version is stored

barbaraliau reviewed (changes requested):
Looks good! See comments.

barbaraliau reviewed (commented):
submitting some comments while i go through the rest of the PR

barbaraliau commented on README.md:
```diff
@@ -64,40 +65,102 @@ Flags:
+Prerequisites:
+
+* Checkout `celo-monorepo` branch `alfajores` and run `yarn && yarn build`
+* Checkout `celo-blockchain` tag `v1.1.2` (NOTE: check that this matches the version specified in the `rosetta` `go.mod` file) and `make all`
```
Where does this specify the version? When I go to `rosetta/go.mod` it doesn't show `1.1.2`

```
// Use this to use external build
replace github.com/ethereum/go-ethereum => github.com/celo-org/celo-blockchain v0.0.0-20200612100840-bf2ba25426f9
```

eelanagaraj replied:
are you on the right version locally? should look like this on master (and for me on this branch locally) https://github.com/celo-org/rosetta/blob/4405b7a0c20c9f060d78e211558154d160f193a6/go.mod#L26

barbaraliau replied:
I thought I was on 0.7.6 but i checked out back to master and then back to 0.7.6 and it's showing 1.1.2 now. Might have fatfingered checking out an older tag.

eelanagaraj commented on README.md:
```diff
@@ -64,40 +65,102 @@ Flags:
+### Version 1: Running from `rosetta` source code
 
-Within the docker image, we pack `rosetta` binary and also `geth` binary from celo-blockchain. Rosetta will run both.
+You will need the following three repositories cloned locally: `rosetta` (this repo), [`celo-monorepo`](https://github.com/celo-org/celo-monorepo), [`celo-blockchain`](https://github.com/celo-org/celo-blockchain). You also need: `go >= 1.14`, `rust >= 1.41.0` (blockchain dependency)
```
can be anywhere -- you set the paths in an ENV var later. If you can think of a clearer way of wording that (without repeating that the paths will be set later), lmk!

barbaraliau replied:
Ok makes sense. We can leave it for now.
Suggest making each repo a bullet point so it's more clear what is what

eelanagaraj replied:
✅ split into two sets of bullets, also deleted the repeated ones under dev setup (and pointed to this one place)

barbaraliau commented on README.md:
```diff
@@ -64,40 +65,102 @@ Flags:
+Prerequisites:
+
+* Checkout `celo-monorepo` branch `alfajores` and run `yarn && yarn build`
+* Checkout `celo-blockchain` tag `v1.1.2` (NOTE: check that this matches the version specified in the `rosetta` `go.mod` file) and `make all`
```
Add command to checkout (i always forget how to checkout tags so this is just to save someone time from looking it up)
```sh
$ git fetch --all --tags
$ git checkout tags/v1.1.2 -b v1.1.2-branch
```

eelanagaraj replied:
✅ added a one-liner for this

barbaraliau commented on README.md:
```diff
@@ -64,40 +65,102 @@ Flags:
+
+* Checkout `celo-monorepo` branch `alfajores` and run `yarn && yarn build`
+* Checkout `celo-blockchain` tag `v1.1.2` (NOTE: check that this matches the version specified in the `rosetta` `go.mod` file) and `make all`
+* Export paths to `celo-monorepo` and `celo-blockchain` as `CELO_MONOREPO_PATH` and `CELO_BLOCKCHAIN_PATH` respectively (can be paths relative to `rosetta` repo)
```
In your bash profile:
```bash
export CELO_MONOREPO_PATH=path/to/celo-monorepo/from/$HOME
export CELO_BLOCKCHAIN_PATH=or/path/to/celo-blockchain/relative/to/rosetta
```

eelanagaraj replied:
✅ added lines for this

barbaraliau commented on README.md:
```diff
@@ -64,40 +65,102 @@ Flags:
+* Checkout `celo-monorepo` branch `alfajores` and run `yarn && yarn build`
+* Checkout `celo-blockchain` tag `v1.1.2` (NOTE: check that this matches the version specified in the `rosetta` `go.mod` file) and `make all`
+* Export paths to `celo-monorepo` and `celo-blockchain` as `CELO_MONOREPO_PATH` and `CELO_BLOCKCHAIN_PATH` respectively (can be paths relative to `rosetta` repo)
+* Checkout `rosetta` tag `v0.7.6` (or latest released tag) and `make gen-contracts && make all`
```
Checkout `rosetta` tag `v0.7.6` (or latest released tag)

```sh
$ git fetch --all --tags
$ git checkout tags/v0.7.6 -b v0.7.6-branch
```

eelanagaraj replied:
✅ added a line for this

barbaraliau reviewed (commented):
more comments

barbaraliau commented on README.md:
```diff
@@ -64,40 +65,115 @@ Flags:
+
+* Checkout `rosetta` tag `v0.7.6` (`git fetch --all && git checkout v0.7.6`) (or latest released tag) and `make gen-contracts && make all`
+* Run `make alfajores-env` to create an empty datadir with the genesis block (only needs to be run the first time, upon initializing the service)
+
```
Suggest adding the following:

You should see output similar to this:

```sh
mkdir -p ./envs/alfajores
curl 'https://storage.googleapis.com/genesis_blocks/alfajores' > ./envs/alfajores/genesis.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:-100 12600  100 12600    0     0  36842      0 --:--:-- --:--:-- --:--:-- 36842
```

eelanagaraj replied:
✅ added for `rc1` as well

barbaraliau commented on README.md:
````diff
@@ -64,40 +65,115 @@ Flags:
+  --geth.staticnodes "enode://05977f6b7d3e16a99d27b714f8a029a006e41ec7732167d373dd920d31f72b3a1776650798d8763560854369d36867e9564dad13b4b60a90c347feeb491d83a9@34.83.42.50:30303" \
+  --datadir "./envs/alfajores"
+```
 
````
Add:

Your system will prompt you to Deny or Allow the application to accept incoming network connections.

You should start to see continuous output similar to this:

```sh
INFO [01-28|14:09:03.434] Press CTRL-C to stop the process
INFO [01-28|14:09:03.440] Running geth init                        service=geth
INFO [01-28|14:09:04.104] Geth Already initialized... skipping init service=geth
geth --networkid 44787 --nousb --rpc --rpcaddr 127.0.0.1 --rpcport 8545 --rpcvhosts localhost --syncmode full --gcmode archive --rpcapi eth,net,web3,debug,admin,personal --ipcpath <YourPathToRosetta>/rosetta/envs/alfajores/celo/geth.ipc --light.serve 0 --light.maxpeers 0 --maxpeers 1100 --consoleformat term
INFO [01-28|14:09:05.110] Detected Chain Parameters                chainId=44787 epochSize=17280
INFO [01-28|14:09:05.120] Starting httpServer                      listen_address=:8080
INFO [01-28|14:09:05.120] Resuming operation from last persisted  block srv=celo-monitor block=0
INFO [01-28|14:09:05.121] SubscriptionFetchMode:Start              srv=celo-monitor pipe=header_listener start=1

...

INFO [01-28|14:09:25.731] Stored 1000 blocks                       srv=celo-monitor pipe=persister       block=1000 registryUpdates=0
INFO [01-28|14:09:34.101] Stored 1000 blocks                       srv=celo-monitor pipe=persister       block=2000 registryUpdates=0
INFO [01-28|14:09:42.625] Stored 1000 blocks                       srv=celo-monitor pipe=persister       block=3000 registryUpdates=0
INFO [01-28|14:09:50.936] Stored 1000 blocks                       srv=celo-monitor pipe=persister       block=4000 registryUpdates=0
INFO [01-28|14:09:59.188] Stored 1000 blocks                       srv=celo-monitor pipe=persister       block=5000 registryUpdates=0
INFO [01-28|14:10:07.479] Stored 1000 blocks                       srv=celo-monitor pipe=persister       block=6000 registryUpdates=0

...

```

eelanagaraj replied:
✅

barbaraliau commented on README.md:
```diff
@@ -64,40 +65,115 @@ Flags:
+### Version 2: Running Rosetta Docker Image
+
+Rosetta is released as a docker image: `us.gcr.io/celo-testnet/rosetta`. All versions can be found on the [registry page](https://us.gcr.io/celo-testnet/rosetta). Within the docker image, we pack the `rosetta` binary and also the `geth` binary from `celo-blockchain`. Rosetta will run both.
+
```
#### Prerequisites:
* [Install](https://docs.docker.com/get-docker/) and run `docker`

eelanagaraj replied:
✅ and added the version that i've tested with

barbaraliau commented on README.md:
```diff
@@ -129,18 +205,14 @@ For a code resource, please see the [examples](./examples/airgap/main.go).
-  * golangci To run linter (check https://github.com/golangci/golangci-lint#install )
+In addition to the dependencies listed above under the instructions for running from `rosetta` source code, you also need:
+
+* openapi-generator To re-generate rpc scaffold ([install link](https://openapi-generator.tech))
```
nit: wrap `openapi-generator` with code formatting

barbaraliau replied:
Does it matter what version?

eelanagaraj replied:
tbh less familiar with this part, a holdover from the previous README (want to keep it around unless we know we can delete). ✅ added formatting

barbaraliau commented on README.md:
```diff
@@ -157,76 +229,27 @@ Rosetta requires a few Celo Core Contracts
-
 ## How to run rosetta-cli-checks
+
 Install the [`rosetta-cli`](https://github.com/coinbase/rosetta-cli) according to the instructions. Current testing has been done with `v0.5.16`.
```
Version `v0.5.16` of rosetta-cli?

eelanagaraj replied:
✅ yes, clarified that a bit

barbaraliau commented on README.md:
```diff
@@ -157,76 +229,27 @@ Rosetta requires a few Celo Core Contracts
-
 ## How to run rosetta-cli-checks
+
 Install the [`rosetta-cli`](https://github.com/coinbase/rosetta-cli) according to the instructions. Current testing has been done with `v0.5.16`.
```
Make this a bullet point. I glossed over it as a to-do.

eelanagaraj replied:
✅

barbaraliau commented on README.md:
````diff
@@ -157,76 +229,27 @@ Rosetta requires a few Celo Core Contracts
+
+```sh
 # alfajores; specify construction or data
 rosetta-cli check:construction --configuration-file PATH/TO/rosetta/rosetta-cli-conf/testnet/cli-config.json
````
I had to run `./bin/rosetta-cli check:construction ...` to get this to run (from the directory I installed it), so recommending installing `rosetta-cli` into `/usr/local/bin` instead of where the command is run would be better. as well as adding instructions to update the command to point to the right path for the `cli-config.json` file

* Install [`rosetta-cli`](https://github.com/coinbase/rosetta-cli)

If on Mac, it is recommended to install `rosetta-cli` into `/usr/local/bin`
```sh
curl -sSfL https://raw.githubusercontent.com/coinbase/rosetta-cli/master/scripts/install.sh | sh -s -- -b /usr/local/bin
```

eelanagaraj replied:
I'm hesitant to add the actual install instructions for the cli here since that'll be an additional piece that we need to ensure is kept up to date (with the `rosetta-cli` README), so I'd prefer deferring to their installation instructions for that reason. That said, I'll add a note that installing to `/usr/local/bin` will allow you to use `rosetta-cli` directly on the command line (without the path to it)

barbaraliau commented on README.md:
````diff
@@ -157,76 +229,27 @@ Rosetta requires a few Celo Core Contracts
+
+```sh
 # alfajores; specify construction or data
 rosetta-cli check:construction --configuration-file PATH/TO/rosetta/rosetta-cli-conf/testnet/cli-config.json
````
I'm actually not sure if I can run checks becaus eI just get a bunch of these
```sh
2021/01/28 14:48:08 waiting for implementation to reach tip before testing...
[MEMORY] Heap: 1103.278084MB Stack: 0.687500MB System: 1260.756973MB GCs: 5
[STATS] Transactions Confirmed: 0 (Created: 0, In Progress: 0, Stale: 0, Failed: 0) Addresses Created: 0
2021/01/28 14:48:18 waiting for implementation to reach tip before testing...
```

How long does it normally take before it reaches the tip? Should I leave it running for a while? If so, I'd add something about that

eelanagaraj replied:
Yes you're all good! It depends on whether you're starting with a fresh node or one with data, etc. Will add a note about that, but yes if this is the first time it'll take quite a while to reach the tip (syncing a full archive node). Data Checks themselves will take a long time since it needs to go through the entire chain + reconcile balances.

eelanagaraj replied:
✅ added a note about this

barbaraliau reviewed (commented):
I've tested almost everything and it looks like everything is g2g except I was getting build errors in the `celo-monorepo` for `rc1`. I'm going to assume that everything would work as expected though.

Actually, would this just be running on `mainnet` instead? Was the `rc1` branch written before the launch of `mainnet`? If. so, then would any `rc1` references need to be updated to `mainnet`/`master`?

barbaraliau reviewed (approved):
Oh but also either way I think these instructions can be merged with what is here and any other changes can be done in a different PR if you want. I'll leave it up to you to decide.

eelanagaraj commented:
> I've tested almost everything and it looks like everything is g2g except I was getting build errors in the `celo-monorepo` for `rc1`. I'm going to assume that everything would work as expected though.
> 
> Actually, would this just be running on `mainnet` instead? Was the `rc1` branch written before the launch of `mainnet`? If. so, then would any `rc1` references need to be updated to `mainnet`/`master`?

@barbaraliau I believe `mainnet` == `rc1`. That said, for `monorepo`, this part is actually a bit more lax for now, as it should only affect the generated contract wrappers (and since there haven't been any relevant breaking changes, it should also even be fine to build a later version of `master` as long as it's pre-breaking changes). (This is something we'll have to address anyways when dealing with contract upgrades, but i'd suggest we punt it until that comes)

eelanagaraj commented:
@barbaraliau addressed your comments, I think it makes sense to push so it'll be easier to maintain + make sure it doesn't get out of date + need to be updated again. In the future, we can also add troubleshooting sections and such, esp. once we see what usage is like of the package. Thanks for your comments!
